Federation performance and bug fixes

hyfen opened the pull request:
Hello RocketChat! @ndroo and I have been working on bug fixes and performance improvements for federated rooms. This PR gathers a bunch of our changes (we can split up into individual PRs if preferred).

A summary of our changes: 
- memoize DNS responses so the checks aren't being done every time a new event arrives
- brings in missing behaviour for threaded messages for federated messages
- improves performance of dispatching system by only sending the same events to a domain once
- fixes a missing import in `app/federation/server/hooks/afterUnsetReaction.js`
- makes the max number of federated domains configurable through an environment variable
- reduces number of times we have to run `federationServers.refreshServers();`

In our repo, we also cherry-picked the unmerged #16430 which is working well.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=17504) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=17504) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/Rocket.Chat?pullRequest=17504) it.</sub>

sampaiodiego commented on app/federation/server/normalizers/room.js:
```diff
@@ -78,6 +80,8 @@ const normalizeRoom = (originalResource, users) => {
 		}
 	}
 
+	domains = _.uniq(domains);
```
```suggestion
	domains = [...new Set(domains)];
```

sampaiodiego commented on app/federation/server/hooks/afterRemoveFromRoom.js:
```diff
@@ -21,7 +23,7 @@ async function afterRemoveFromRoom(involvedUsers, room) {
 	try {
 		// Get the domains after removal
-		const domainsAfterRemoval = users.map((u) => u.federation.origin);
+		const domainsAfterRemoval = _.uniq(users.map((u) => u.federation.origin));
```
```suggestion
		const domainsAfterRemoval = [...new Set(users.map((u) => u.federation.origin))];
```

sampaiodiego commented on app/federation/server/hooks/afterRemoveFromRoom.js:
```diff
@@ -30,7 +32,7 @@ async function afterRemoveFromRoom(involvedUsers, room) {
 
 		// Get the users domains
-		const domainsBeforeRemoval = usersBeforeRemoval.map((u) => u.federation.origin);
+		const domainsBeforeRemoval = _.uniq(usersBeforeRemoval.map((u) => u.federation.origin));
```
```suggestion
		const domainsBeforeRemoval = [...new Set(usersBeforeRemoval.map((u) => u.federation.origin))];
```

sampaiodiego commented on app/federation/server/hooks/afterLeaveRoom.js:
```diff
@@ -28,7 +30,7 @@ async function afterLeaveRoom(user, room) {
 
 		// Get the users domains
-		const domainsBeforeLeft = usersBeforeLeave.map((u) => u.federation.origin);
+		const domainsBeforeLeft = _.uniq(usersBeforeLeave.map((u) => u.federation.origin));
```
```suggestion
		const domainsBeforeLeft = [...new Set(usersBeforeLeave.map((u) => u.federation.origin))];
```

sampaiodiego commented on app/federation/server/hooks/afterLeaveRoom.js:
```diff
@@ -19,7 +21,7 @@ async function afterLeaveRoom(user, room) {
 	try {
 		// Get the domains after leave
-		const domainsAfterLeave = users.map((u) => u.federation.origin);
+		const domainsAfterLeave = _.uniq(users.map((u) => u.federation.origin));
```
```suggestion
		const domainsAfterLeave = [...new Set(users.map((u) => u.federation.origin))];
```

sampaiodiego commented on app/federation/server/handler/index.js:
```diff
@@ -65,7 +68,11 @@ export function dispatchEvents(domains, events) {
-	dispatchEvents(domains, [event]);
+
+	//Ensure the domain list is distinct to avoid excessive events
+	const distinctDomains = _.uniq(domains).filter(domain => event.origin);
```
```suggestion
	const distinctDomains = [...new Set(domains).filter(domain => event.origin)];
```

sampaiodiego commented on app/federation/server/handler/index.js:
```diff
@@ -55,6 +56,8 @@ export function dispatchEvents(domains, events) {
 		throw disabled('client.dispatchEvents');
 	}
 
+	domains = _.uniq(domains);
```
```suggestion
	domains = [...new Set(domains)];
```

hyfen commented:
@sampaiodiego Switched to using `Set` and fixed the linting issues. 

Also the linting pointed me to a bug that I had introduced in an earlier refactoring. Fixed that here: https://github.com/RocketChat/Rocket.Chat/pull/17504/commits/7e225daf7cc17c44232800228ac5f987b770f687#diff-2ddd532f087a155b46f014ddedcf3102R71

sampaiodiego commented:
> @sampaiodiego Switched to using `Set` and fixed the linting issues.
> 
> Also the linting pointed me to a bug that I had introduced in an earlier refactoring. Fixed that here: [7e225da#diff-2ddd532f087a155b46f014ddedcf3102R71](https://github.com/RocketChat/Rocket.Chat/commit/7e225daf7cc17c44232800228ac5f987b770f687#diff-2ddd532f087a155b46f014ddedcf3102R71)

oops, that was my fault, bad suggestion :/

but thanks again.. 👍

sampaiodiego commented:
thanks @ndroo for the explanation.. I see your point, but that `await` was not helping to prevent that to happen. `await` is effective only when using Promises, that is not the case for `Messages.insert`.. actually, any model action is already synchronous due to Meteor's Fibers (except for raw models, that doesn't use Meteor's collections thus doesn't rely on Fibers), that's why I removed the `await`, the behaviour should stay the same.

> The challenge is that due to the way the federation system works, if a server is backlogged in past events, it might get sent the same event multiple times. When this occurs, if you try and insert the message a second time, it'll fatally fail with a duplicate key error. 

@alansikora any way to prevent this from happening?

ndroo commented:
Hmm ok yeah, i'm honestly very new to all this but the situation seemed to have been resolved when i added it. Perhaps I didn't have a reliable enough test though?

I guess using upsert v's insert could also mitigate the issue.

sampaiodiego commented:
maybe you can share how you tested with us so we can try reproduce it too and think on other ways to fix it? I don't want to merge a buggy code by no means, specially one that I have changed without testing properly =)

using `upsert` could do the tricky, but just like the try/catch it can possibly hide an unwanted behaviour.. I had a quick discussion with @alansikora and he mentioned [this](https://github.com/RocketChat/Rocket.Chat/pull/17504/files#diff-4e37c2f6d63303fdc8a6283be65bfc6cR257) try/catch might not be there.. as if some error happens we maybe shouldn't continue.

ndroo commented:
Yeah, so i see the point, it should probably check for just a MySQL
duplicate key. That issue was breaking federation on a regular basis for me
in testing. It’s pretty simple to expose the crash, connect two servers,
and mashing the keyboard on both at the same time make a bunch of messages.
You’ll get a crash due to duplicate keys  (same event being processed
multiple times) in a matter of seconds.

On Mon, May 4, 2020 at 18:04 Diego Sampaio <notifications@github.com> wrote:

> maybe you can share how you tested with us so we can try reproduce it too
> and think on other ways to fix it? I don't want to merge a buggy code by no
> means, specially one that I have changed without testing properly =)
>
> using upsert could do the tricky, but just like the try/catch it can
> possibly hide an unwanted behaviour.. I had a quick discussion with
> @alansikora <https://github.com/alansikora> and he mentioned this
> <https://github.com/RocketChat/Rocket.Chat/pull/17504/files#diff-4e37c2f6d63303fdc8a6283be65bfc6cR257>
> try/catch might not be there.. as if some error happens we maybe shouldn't
> continue.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/RocketChat/Rocket.Chat/pull/17504#issuecomment-623732457>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFNDXJ7T3B2U7YIWDWIA7DRP43XLANCNFSM4MWWFECQ>
> .
>
-- 
Andrew McGrath

Ph. 647-225-4909
Timezone: Toronto/EST (UTC -4)
